### ISSUE TITLE
Bug w/Repro Repo: Custom node and plugin causes TextNode to strip trailing whitespace on importDOM

### ISSUE THREAD

**Issue by JakeHadley** [facebook/lexical#4954]
tl;dr: The issue is when the $generateNodesFromDOM function runs, the Lexical TextNode has it's trailing whitespace stripped out when followed by my custom node.

I'm trying to build out my own extended formatting text node. Currently this example only has a bold option, but I'm looking to utilize this to provide different font-weights, among other things. I have steps recorded via Windows Steps Recorder, but I essentially created a few plugins to create a custom text node, load existing html that is formatted for our requirements, and a toolbar. When I format my custom node via one of the buttons on the toolbar, it adds a class to the node. On exportDOM, I create my own custom HTML tag for our internal storage. On importDOM, I translate that custom tag to be a span with the correct class. 

Lexical version: ^0.12.0

## Steps To Reproduce

1. Hit "Primary" button to enable the editor
2. Highlight "something" and hit the bold icon button. Text is bolded
3. Hit the check button icon to close the editor. Text is still bolded.
4. Hit "Primary" button again to enable the editor. Unbolded text now has it's trailing space removed

![lexical-repro](https://github.com/facebook/lexical/assets/35585841/f03963f7-b541-4e76-bf64-2e16ca7f7c38)

It's a nextjs app. Clone the repo, and from the root folder run `npm run dev`.
Link to code example: https://github.com/JakeHadley/lexical-test

## The current behavior
Lexical TextNode whitespace is stripped out on importDOM

## The expected behavior
The text should appear in the editor the same way as it appears outside the editor

**Comment by acywatson** [comment]
@zurfyx I think this is related to the whitespace changes. Second issue about this, I believe.

**Comment by acywatson** [comment]
Here's the other issue: https://github.com/facebook/lexical/issues/4919

**Comment by JakeHadley** [comment]
@acywatson I'm considering opening this back up, but I wanted to confirm first. I have this same issue happening again, but it's with a custom bare bones text node that I'm using to override TextNode so that I can control the export of the html that happens from `generateHtmlFromNodes`. importDOM is getting called in the node, but I guess I don't know for sure if the super call actually works the way it should or not. I'm essentially trying to just get the text exported from the regular text nodes, which is why I'm overriding and exporting `{ element: new Text(this.__text) }` from `exportDOM`.

This is the node, and it's having the whitespace issue again:
```
import { DOMConversionMap, DOMExportOutput, SerializedTextNode, Spread, TextNode } from 'lexical'

export type SerializedCustomTextNode = Spread<unknown, SerializedTextNode>
const CUSTOMTEXTTYPE = 'custom-text'

export class CustomTextNode extends TextNode {
  public static clone(node: CustomTextNode) {
    return new CustomTextNode(node.__text)
  }

  public static getType(): string {
    return CUSTOMTEXTTYPE
  }

  public static importDOM(): DOMConversionMap | null {
    return super.importDOM()
  }

  public static importJSON(serializedNode: SerializedTextNode): TextNode {
    return super.importJSON(serializedNode)
  }

  public exportDOM(): DOMExportOutput {
    return { element: new Text(this.__text) }
  }

  public exportJSON(): SerializedTextNode {
    return super.exportJSON()
  }
}

export function $createCustomTextNode(text: string): CustomTextNode {
  return new CustomTextNode(text)
}
```
And in the editor config:
```
const initialConfig = {
    namespace: 'MyEditor',
    theme,
    onError,
    nodes: [
      FormattedTextNode,
      LinkTextNode,
      CustomTextNode,
      {
        replace: TextNode,
        with: (node: TextNode) => {
          return new CustomTextNode(node.__text)
        }
      }
    ]
  }
  ```
